File: src/k_functionParameters/functionParameters.java
package k_functionParameters;

import java.util.function.Consumer;
import java.util.function.Function;

public class functionParameters {
    public static void main(String[] args) {
        Function<String, Integer> function1 = arg -> arg.length();
        Function<Void, Integer> function2 = arg -> 42;
        Consumer<Void> function3 = arg -> System.out.println("Hello");
        Consumer<Object> function4 = arg -> System.out.println(arg);

        callFunction(function1, function4);
    }

    public static void callFunction(Function<String, Integer> lengthFunction, Consumer<Object> printFunction) {
        int length = lengthFunction.apply("Hello");
        printFunction.accept(length);
    }
}

File: src/k_functionParameters/functionParameters.kt
package k_functionParameters

fun main(args: Array<String>) {
    val function1: (String) -> Int = { arg: String -> arg.length }
    val function2: () -> Int = { 42 }
    val function3: () -> Unit = { println("Hello") }
    val function4: (Any) -> Unit = { arg: Any -> println(arg) }

    callFunction(lengthFunction = function1, printFunction = function4)
}

fun callFunction(lengthFunction: (String) -> Int, printFunction: (Any) -> Unit) {
    val length = lengthFunction("Hello")
    printFunction(length)
}
